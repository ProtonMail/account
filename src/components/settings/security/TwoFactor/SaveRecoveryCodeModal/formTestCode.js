import { connect } from 'unistore/full/preact';
import settingsActions from '../../../../../actions/settings';

import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';


import styles from './index.css';

const renderInfo = (result) => {
    if (typeof result === 'undefined') {
        // result is undefined until the input is lower than 6 characters.
        return (<p>Your recovery code will not be erased</p>);
    }

    return <p>{result ? '✓ Test succeeded' : '⚠ Please test your recovery code to proceed'}</p>;
};

/**
 * Modal Form to test that a given code is one of the new recovery code.
 * @param {Object} props
 * @param {Function} props.onSubmit - triggers the next step.
 * @param {Function} props.onCancel - triggers the previous step.
 * @returns {ModalWrapper}
 */
export const FormTestCode = ({
    onSubmit,
    onCancel,
    settings: { reset2FARecoveryCodes: { result, response = {} } },
    reset2FARecoveryCodesCheckNewCodeAction
}) => {
    const model = { code: response.code };

    return (
        <ModalWrapper
            onSubmit={(e) => {
                e.preventDefault();
                onSubmit();
            }}
            onReset={(e) => {
                e.preventDefault();
                onCancel();
            }}
        >
            <ModalContent>
                <p class={styles.description}>
                    Test your recovery codes by entering one of your codes below. If you did not save your recovery
                    codes, go back and save them.
                </p>
                <div class="form-row">
                    <label htmlFor="verifyCode">Input your code</label>
                    <div>
                        <input
                            onInput={({ target: { value: code } }) => {
                                model.code = code;
                                if (code.length === 8) {
                                    reset2FARecoveryCodesCheckNewCodeAction(model.code);
                                }
                            }}
                            required={true}
                            value={model.code}
                            type="code"
                            id="verifyCode"
                            placeholder="Code"
                            disabled={!!result}
                            autoFocus={true}
                        />
                    </div>
                </div>
                <div class={styles.result}>
                    {renderInfo(result)}
                </div>
            </ModalContent>
            <ModalFooter>
                <button type="reset" value="Reset">
                    Back
                </button>
                <button type="submit" value="Submit" disabled={!result}>
                    Finish
                </button>
            </ModalFooter>
        </ModalWrapper>
    );
};

export default connect('settings', settingsActions)(FormTestCode);
