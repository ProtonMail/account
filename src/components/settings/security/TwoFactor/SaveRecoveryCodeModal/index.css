.list {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    align-self: stretch;
    padding: 0 5%;
}

.item {
    width: 15%;
    margin-left: 10%;
}

.code {
}

.actions {
    display: flex;
    justify-content: space-around;
    align-self: stretch;
    text-align: left;
}

.description {
    flex: 2;
}
.result{
    flex: 2;
}
