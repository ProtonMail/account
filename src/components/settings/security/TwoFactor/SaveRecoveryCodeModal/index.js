import Presentation from './presentation';
import TestCode from './testCode';

export default [
    {
        title: 'Save your recovery codes',
        component: (params, onSubmit, onCancel) => (
            <Presentation params={params} onSubmit={onSubmit} onCancel={onCancel} />
        )
    },
    {
        title: 'Test your recovery codes',
        component: (params, onSubmit, onCancel) => <TestCode params={params} onSubmit={onSubmit} onCancel={onCancel} />
    }
];
