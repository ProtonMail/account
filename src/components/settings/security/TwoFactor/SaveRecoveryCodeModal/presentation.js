import styles from './index.css';
import { styles as ModalStyles } from '../../../../ui/Modal';
import { saveAs } from 'file-saver';

import { CopyToClipboard } from 'react-copy-to-clipboard';

/**
 * generates a TXT document, and download it from the browser.
 * @param {String[]} codes - the list of codes.
 */
function downloadClicked ( codes ) {
    const blob = new Blob([ codes.join('\r\n') ], { type: 'text/plain;charset=utf-8' });
    saveAs(blob, 'proton-recovery-codes.txt');
}

/**
 * Display a list of codes.
 * @param {Object} params - the params.
 * @param {String[]} params.codes - the list of codes.
 * @param {Function} onSubmit - triggered when next is pressed.
 * @param {Function} onCancel - triggered when previous is pressed.
 * @return {Component}
 */
export default ( { params: { codes }, onSubmit, onCancel } ) => {
    if (!codes) {
        codes = [
            'd1e4822e',
            '2d83d85b',
            '717ebed4',
            'a9ffee38',
            'f34ebdf7',
            'd1321481',
            '2572f6cd',
            '39bd63b3',
            'c8016641',
            'a9b4cf9d',
            '8f475f77',
            '123d33b1'
        ];
    }

    return (
        <form
            class={ModalStyles.wrapper}
            onSubmit={( e ) => {
                e.preventDefault();
                onSubmit({ codes });
            }}
            onReset={( e ) => {
                e.preventDefault();
                onCancel({ codes });
            }}
        >
            <div class={ModalStyles.content}>
                <p>
                    Please keep your recovery codes in a safe place. Otherwise, you can permanently lose access to your
                    account if you loose your 2FA device
                </p>
                <p>Each recovery code can only be used once</p>
                <ol class={styles[ 'CodePresentation-List' ]}>
                    {codes.map(( code ) => (
                        <li>
                            <pre>{code}</pre>
                        </li>
                    ))}
                </ol>
                <div class={styles[ 'CodePresentation-Actions' ]}>
                    <a href="#" onClick={() => downloadClicked(codes)}>
                        DOWNLOAD CODES
                    </a>
                    <CopyToClipboard text={codes.join('\n')}>
                        <a href={'#'}>COPY CODES </a>
                    </CopyToClipboard>
                </div>
            </div>
            <div class={ModalStyles.footer}>
                <button type="reset" value="Reset">
                    Back
                </button>
                <button type="submit" value="Submit">
                    Next
                </button>
            </div>
        </form>
    );
};
