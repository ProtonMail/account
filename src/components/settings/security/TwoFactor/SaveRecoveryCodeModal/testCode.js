import { h, Component } from 'preact';
import { styles as ModalStyles } from '../../../../ui/Modal';
import styles from './index.css';

/**
 * Asks the user to give on of the recovery codes to be able to continue.
 */
export default class TestCodesModal extends Component {

    /**
     * verifies the code is valid.
     * @param {String} code - the code to be checked.
     * @return {boolean}
     */
    isValid ( code ) {
        return this.props.params.codes.indexOf(code) > -1;
    }

    /**
     * Verfies the code each time it is updated by the user. If the code is not 8 chars, it will not be checked.
     * @param {Event} e - the changed event.
     */
    onCodeChange ( e ) {
        const code = e.target.value;
        let valid = 0;
        if (code.length === 8) {
            valid = 1 + this.isValid(code);
        }
        this.setState({ code: e.target.value, valid });
    }

    /**
     * @constructor
     * @param {Object} props
     * @param {Object} props.params
     * @param {String[]} props.params.codes - the list of recovery codes.
     */
    constructor ( props ) {
        super(props);
        this.state = {
            code: '',
            valid: 0
        };
    }

    /**
     * Render the code section: either an input, if the code was not validated, or a span telling the code is validated.
     * @return {Component}
     */
    renderCode () {
        if (this.state.valid === 2) {
            return <span>v Test succeeded</span>;
        }
        return (
            <input
                onInput={this.onCodeChange.bind(this)}
                value={this.state.code}
                style={{ width: '100%' }}
                type="code"
                id="verifyCode"
                placeholder="Code"
            />
        );
    }

    render () {
        return (
            <form
                class={ModalStyles.wrapper}
                onSubmit={( e ) => {
                    e.preventDefault();
                    this.props.onSubmit({});
                }}
                onReset={( e ) => {
                    e.preventDefault();
                    this.props.onCancel({});
                }}
            >
                <div class={ModalStyles.content}>
                    <p style={{ flex: 2 }}>
                        Test your recovery codes by entering one of your codes below. If you did not save your recovery
                        codes, go back and save them.
                    </p>
                    <div class={styles[ 'TestCodeForm-InputContainer' ]}>
                        <span>Input your code</span>
                        <div>{this.renderCode()}</div>
                    </div>
                    {this.state.valid === 1 && (
                        <div style={{ flex: 1 }}>
                            <p>! Please test one of your recovery code</p>
                        </div>
                    )}
                </div>
                <div class={ModalStyles.footer}>
                    <button type="reset" value="Reset">
                        Back
                    </button>
                    <button type="submit" value="Submit" disabled={this.state.valid !== 2}>
                        Finish
                    </button>
                </div>
            </form>
        );
    }
}
