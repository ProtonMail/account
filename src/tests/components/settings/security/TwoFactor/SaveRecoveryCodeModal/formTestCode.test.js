import { deep } from 'preact-render-spy';
import render from 'preact-render-to-string';

import { FormTestCode } from '../../../../../../components/settings/security/TwoFactor/SaveRecoveryCodeModal/formTestCode';

const renderFormTestCode = (props = {}, { code = '', result = undefined } = {}) => {
    const response = code ? { code } : undefined;
    return (
        <FormTestCode
            {...props}
            settings={{
                reset2FARecoveryCodes: { result, response }
            }}
        />
    );
};

describe('test for SaveRecoveryCodeModal SaveCode step', () => {
    test('initial display', () => {
        expect(render(renderFormTestCode())).toMatchSnapshot();
    });

    test('ongoing display', () => {
        expect(render(renderFormTestCode({}, { code: '945dc' }))).toMatchSnapshot();
    });

    test('success display', () => {
        expect(render(renderFormTestCode({}, { result: true }))).toMatchSnapshot();
    });

    test('failure display', () => {
        expect(render(renderFormTestCode({}, { result: false, code: '945dc945dc' }))).toMatchSnapshot();
    });

    test('on input test', () => {
        const reset2FARecoveryCodesCheckNewCodeAction = jest.fn();
        const context = deep(
            renderFormTestCode(
                { reset2FARecoveryCodesCheckNewCodeAction },
                {
                    result: false,
                    code: '945dc'
                }
            )
        );

        context.find('input').simulate('input', { target: { value: '945dcc' } });
        expect(reset2FARecoveryCodesCheckNewCodeAction).not.toHaveBeenCalled();

        context.find('input').simulate('input', { target: { value: '12345678' } });
        expect(reset2FARecoveryCodesCheckNewCodeAction).toHaveBeenCalledWith('12345678');
    });

    test('submit and cancel', () => {
        const onSubmit = jest.fn();
        const onCancel = jest.fn();
        const context = deep(
            renderFormTestCode(
                { onSubmit, onCancel },
                {
                    result: false,
                    code: '945dc'
                }
            )
        );

        const event = { preventDefault: () => undefined };

        context.find('form').simulate('submit', event);
        expect(onSubmit).toHaveBeenCalled();
        expect(onCancel).not.toHaveBeenCalled();

        context.find('form').simulate('reset', event);
        expect(onSubmit).toHaveBeenCalledTimes(1); // the submit event
        expect(onCancel).toHaveBeenCalled(); // only once
    });
});
