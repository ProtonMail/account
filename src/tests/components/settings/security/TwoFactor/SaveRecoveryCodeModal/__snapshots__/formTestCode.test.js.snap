// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for SaveRecoveryCodeModal SaveCode step failure display 1`] = `
<form class="wrapper">
  <div class="content ">
    <p class="description">Test your recovery codes by entering one of your codes below. If you did
      not save your recovery codes, go back and save them.</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input required value="945dc945dc" type="code" id="verifyCode" placeholder="Code"
        autoFocus />
      </div>
    </div>
    <div class="result">
      <p>⚠ Please test your recovery code to proceed</p>
    </div>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit" disabled>Finish</button>
  </div>
</form>
`;

exports[`test for SaveRecoveryCodeModal SaveCode step initial display 1`] = `
<form class="wrapper">
  <div class="content ">
    <p class="description">Test your recovery codes by entering one of your codes below. If you did
      not save your recovery codes, go back and save them.</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input required type="code" id="verifyCode" placeholder="Code" autoFocus
        />
      </div>
    </div>
    <div class="result">
      <p>Your recovery code will not be erased</p>
    </div>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit" disabled>Finish</button>
  </div>
</form>
`;

exports[`test for SaveRecoveryCodeModal SaveCode step ongoing display 1`] = `
<form class="wrapper">
  <div class="content ">
    <p class="description">Test your recovery codes by entering one of your codes below. If you did
      not save your recovery codes, go back and save them.</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input required value="945dc" type="code" id="verifyCode" placeholder="Code"
        autoFocus />
      </div>
    </div>
    <div class="result">
      <p>Your recovery code will not be erased</p>
    </div>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit" disabled>Finish</button>
  </div>
</form>
`;

exports[`test for SaveRecoveryCodeModal SaveCode step success display 1`] = `
<form class="wrapper">
  <div class="content ">
    <p class="description">Test your recovery codes by entering one of your codes below. If you did
      not save your recovery codes, go back and save them.</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input required type="code" id="verifyCode" placeholder="Code" disabled
        autoFocus />
      </div>
    </div>
    <div class="result">
      <p>✓ Test succeeded</p>
    </div>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Finish</button>
  </div>
</form>
`;
